pci/src/vfio.rs

```diff
@@ -861,63 +847,82 @@ impl PciDevice for VfioPciDevice {
             // By default, the region type is 32 bits memory BAR.
             let mut region_type = PciBarRegionType::Memory32BitRegion;
 
+            // To get size write all 1s
+            self.vfio_pci_configuration
+                .write_config_dword(0xffff_ffff, bar_offset);
+
+            // And read back BAR value. The device will write zeros for bits it doesn't care about
+            let mut lower = self.vfio_pci_configuration.read_config_dword(bar_offset);
+
             if io_bar {
                 #[cfg(target_arch = "x86_64")]
                 {
                     // IO BAR
                     region_type = PciBarRegionType::IoRegion;
 
-                    // Clear first bit.
-                    lsb_size &= 0xffff_fffc;
+                    // Mask flag bits (lowest 2 for I/O bars)
+                    lower &= !0b11;
+
+                    // BAR is not enabled
+                    if lower == 0 {
+                        bar_id += 1;
+                        continue;
+                    }
+
+                    // Invert bits and add 1 to calculate size
+                    region_size = (!lower + 1) as u64;
```

Comment:
Might be better to create `lower` as u64 from the start, so that we don't have to cast it to u64 in here. I'm worried that if `lower` is created as u32 and its inverted value is `0xffff_ffff`, by adding 1 we're gonna exceed what `lower` can hold.

Reply:
It's always good to think about overflow but in this case it can't happen. The only way to overflow when adding 1 is if the inverted value was `0xffff_ffff` which requires the non-inverted value to be 0 which can't happen as that value is checked earlier and cause the loop to continue.

I don't think using a u64 to store a value (from the config register) which is definitely a u32 is not a good idea from a correctness perspective.

Reply:
:+1: